### PR TITLE
Add PhantomSpawnEvent

### PR DESCRIPTION
**Pull request by BillyGalbreath**

### CODE REVIEW

**Comment by Black-Hole** [comment]
How about extending PreCreatureSpawnEvent with a @Nullable entity that causes the spawn?

**Comment by aikar** [comment]
This event was designed to fire before the entity was created. this makes it slightly inconsistent. 

We could maybe do something like setting a world.spawningPlayer = player and make CreatureSpawnEvent have a .getTriggeringPlayer() ?

I don't know if we need a new event specifically for phantoms like this, and im not really a fan of mixing up the the context the pre event.

**Comment by aikar** [comment]
though my concerns would be resolved if we did not pass the phantom entity and fired it earlier.

**Comment by Black-Hole** [comment]
By "extending" I was talking about adding more functionality to the PreCreatureSpawnEvent. This could be useful for other spawns like SpawnReason.TRAP or SpawnReason.ENDER_PEARL.

How about using a more general "Paper.spawningEntity" instead of "Paper.spawningPlayer"?

File: Spigot-Server-Patches/0351-Add-PhantomSpawnEvent.patch [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+         this.setSize(nbttagcompound.getInt("Size"));
++        this.spawningPlayer = nbttagcompound.getUUID("Paper.spawningPlayer"); // Paper
+     }
+ 
+     public void b(NBTTagCompound nbttagcompound) {
+@@ -136,6 +137,7 @@ public class EntityPhantom extends EntityFlying implements IMonster {
+         nbttagcompound.setInt("AY", this.c.getY());
+         nbttagcompound.setInt("AZ", this.c.getZ());
+         nbttagcompound.setInt("Size", this.getSize());
++        nbttagcompound.setUUID("Paper.spawningPlayer", this.spawningPlayer); // Paper
```
Comment (zachbr) [inline Spigot-Server-Patches/0351-Add-PhantomSpawnEvent.patch]: `Paper.spawningPlayer` -> `Paper.SpawningPlayer`

File: Spigot-Server-Patches/0351-Add-PhantomSpawnEvent.patch [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+diff --git a/src/main/java/net/minecraft/server/EntityPhantom.java b/src/main/java/net/minecraft/server/EntityPhantom.java
+index c509dfe0d..a4b6083ad 100644
+--- a/src/main/java/net/minecraft/server/EntityPhantom.java
++++ b/src/main/java/net/minecraft/server/EntityPhantom.java
+@@ -128,6 +128,7 @@ public class EntityPhantom extends EntityFlying implements IMonster {
+         }
+ 
+         this.setSize(nbttagcompound.getInt("Size"));
++        this.spawningPlayer = nbttagcompound.getUUID("Paper.spawningPlayer"); // Paper
```
Comment (zachbr) [inline Spigot-Server-Patches/0351-Add-PhantomSpawnEvent.patch]: `Paper.spawningPlayer` -> `Paper.SpawningPlayer`

**Review by aikar — CHANGES_REQUESTED** [review]
To make it very clear this is a prespawn event, let's name this PhantomPreSpawnEvent

**Pull request merged**